crates/typst-pdf/src/embed.rs

```diff
@@ -0,0 +1,90 @@
+use crate::catalog::pdf_date;
+use crate::{PdfAConformanceLevel, PdfChunk, WithGlobalRefs};
+use ecow::EcoString;
+use pdf_writer::{Finish, Name, Ref, Str, TextStr};
+use std::collections::HashMap;
+use typst_library::diag::{bail, SourceResult};
+use typst_library::foundations::NativeElement;
+use typst_library::pdf::embed::{Embed, EmbedElem, EmbeddedFileRelationship};
+use typst_syntax::Span;
+
+pub fn write_embedded_files(
+    ctx: &WithGlobalRefs,
+) -> SourceResult<(PdfChunk, HashMap<EcoString, Ref>)> {
+    let mut chunk = PdfChunk::new();
+
+    let elements = ctx.document.introspector.query(&EmbedElem::elem().select());
+    if !elements.is_empty()
+        && Some(PdfAConformanceLevel::A_2) == ctx.options.standards.pdfa
+    {
+        bail!(
+            Span::detached(),
+            "file embeddings are currently only supported for PDF/A-3"
+        );
+    }
+
+    let mut embedded_files = HashMap::default();
+    for elem in elements.iter() {
+        let packed_elem = elem.to_packed::<EmbedElem>().unwrap();
+        let embed = Embed::from_element(packed_elem);
+        embedded_files.insert(embed.name().clone(), embed_file(ctx, &mut chunk, &embed));
+    }
+
+    Ok((chunk, embedded_files))
+}
+
+fn embed_file(ctx: &WithGlobalRefs, chunk: &mut PdfChunk, embed: &Embed) -> Ref {
+    let embedded_file_stream_ref = chunk.alloc.bump();
+    let file_spec_dict_ref = chunk.alloc.bump();
+
+    let length = embed.data().len();
+
+    let mut embedded_file =
+        chunk.embedded_file(embedded_file_stream_ref, embed.data().as_ref());
+    embedded_file.pair(Name(b"Length"), length as i32);
+    if let Some(mime_type) = embed.mime_type() {
+        embedded_file.subtype(Name(mime_type.as_bytes()));
+    }
+    let date = ctx
+        .document
+        .info
+        .date
+        .unwrap_or(ctx.options.timestamp)
+        .and_then(|date| pdf_date(date, ctx.document.info.date.is_auto()));
+    if let Some(pdf_date) = date {
+        embedded_file.params().modification_date(pdf_date).finish();
+    }
+    embedded_file.finish();
+
+    let mut file_spec = chunk.file_spec(file_spec_dict_ref);
+    file_spec
+        .path(Str(embed.path().as_bytes()))
+        .unic_file(TextStr(embed.path().as_str()))
+        .insert(Name(b"EF"))
+        .dict()
+        .pair(Name(b"F"), embedded_file_stream_ref)
+        .pair(Name(b"UF"), embedded_file_stream_ref)
+        .finish();
```

Comment:
This finish here should be removed. It doesn't have any effect, but is not needed and confusing.